### ISSUE TITLE
SIGSEGV in valhalla_service when using synchronized global cache

### ISSUE THREAD

**Issue by purew** [valhalla/valhalla#2122]
I'm seeing segmentation faults in `valhalla_service` when using `global_synchronized_cache` and more than 1 worker thread _and_ issuing requests in parallel:
```
2019/12/20 13:48:29.034697 [ANALYTICS] total_location_distance::0.144637km
2019/12/20 13:48:29.057249 [ANALYTICS] travel_mode::2
Thread 10 "valhalla_servic" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd3fff700 (LWP 2415821)]
0x00005555557738db in valhalla::baldr::GraphReader::GetOpposingEdgeId(valhalla::baldr::GraphId const&, valhalla::baldr::GraphTile const*&) ()
(gdb) 
```
Backtrace
```
(gdb) bt
#0  0x00005555557738db in valhalla::baldr::GraphReader::GetOpposingEdgeId(valhalla::baldr::GraphId const&, valhalla::baldr::GraphTile const*&) ()
#1  0x000055555579ac57 in (anonymous namespace)::bin_handler_t::handle_bin(__gnu_cxx::__normal_iterator<(anonymous namespace)::projector_wrapper*, std::vector<(anonymous namespace)::projector_wrapper, std::allocator<(anonymous namespace)::projector_wrapper> > >, __gnu_cxx::__normal_iterator<(anonymous namespace)::projector_wrapper*, std::vector<(anonymous namespace)::projector_wrapper, std::allocator<(anonymous namespace)::projector_wrapper> > >) ()
#2  0x00005555557a003f in valhalla::loki::Search(std::vector<valhalla::baldr::Location, std::allocator<valhalla::baldr::Location> > const&, valhalla::baldr::GraphReader&, valhalla::sif::DynamicCost const*) ()
#3  0x000055555565fdf6 in valhalla::loki::loki_worker_t::route(valhalla::Api&) ()
#4  0x000055555565730f in valhalla::loki::loki_worker_t::work(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&) ()
#5  0x0000555555658b11 in std::_Function_handler<prime_server::worker_t::result_t (std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()>&), std::_Bind<prime_server::worker_t::result_t (valhalla::loki::loki_worker_t::*(std::reference_wrapper<valhalla::loki::loki_worker_t>, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>))(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&)> >::_M_invoke(std::_Any_data const&, std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*&&, std::function<void ()>&) ()
#6  0x00007ffff7d8f1e6 in prime_server::worker_t::work() () from /usr/lib/libprime_server.so.0
#7  0x0000555555658506 in valhalla::loki::run_service(boost::property_tree::basic_ptree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) ()
#8  0x00007ffff7523ee4 in std::execute_native_thread_routine (__p=0x55555614eb70)
    at /build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
#9  0x00007ffff72d94cf in start_thread () from /usr/lib/libpthread.so.0
#10 0x00007ffff72082d3 in clone () from /usr/lib/libc.so.6
(gdb) 
```

**Comment by PureW** [comment]
With debug symbols:
```
2019/12/20 14:08:48.945539 [ANALYTICS] locations_count::2                                                                                                                            [43/55436]
2019/12/20 14:08:48.945599 [ANALYTICS] costing_type::bicycle                            
2019/12/20 14:08:48.945638 [ANALYTICS] total_location_distance::9.536551km                     
                                                         
Thread 17 "valhalla_servic" received signal SIGSEGV, Segmentation fault.                               
[Switching to Thread 0x7fffc3fff700 (LWP 2450702)]       
0x00005555556af9d2 in valhalla::baldr::DirectedEdge::toll (this=0x7fff5a67dfe8)                  
    at ../valhalla/baldr/directededge.h:258
258         return toll_;                                                                                         
(gdb) bt                                                                                                          
#0  0x00005555556af9d2 in valhalla::baldr::DirectedEdge::toll (this=0x7fff5a67dfe8)     
    at ../valhalla/baldr/directededge.h:258
#1  0x00005555556b0d35 in valhalla::sif::DynamicCost::base_transition_cost (this=0x7fffb045a280,                  
    node=0x7fffb187dc90, edge=0x7fff5a67dfe8, pred=0x7fffb1ad7538, idx=0) at ../valhalla/sif/dynamiccost.h:740    
#2  0x00005555556be4f4 in valhalla::sif::BicycleCost::TransitionCostReverse (this=0x7fffb045a280, idx=0,          
    node=0x7fffb187dc90, pred=0x7fffb1ad7538, edge=0x7fff5a67dfe8) at ../src/sif/bicyclecost.cc:778               
#3  0x000055555597cb37 in valhalla::thor::BidirectionalAStar::ExpandReverseInner (this=0x7fffc3ffe1b0,            
    graphreader=..., pred=..., opp_pred_edge=0x7fff5a67dfe8, nodeinfo=0x7fffb187dc90, pred_idx=26094, meta=...,   
    shortcuts=@0x7fffc3ffccc4: 0, tile=0x7fffb043c710) at ../src/thor/bidirectional_astar.cc:434
#4  0x0000555555977f32 in valhalla::thor::BidirectionalAStar::ExpandReverse (this=0x7fffc3ffe1b0,                 
    graphreader=..., node=..., pred=..., pred_idx=26094, opp_pred_edge=0x7fff5a67dfe8, from_transition=false)
    at ../src/thor/bidirectional_astar.cc:324                                                                     
#5  0x0000555555979138 in valhalla::thor::BidirectionalAStar::GetBestPath (this=0x7fffc3ffe1b0, origin=...,                                                                                    
    destination=..., graphreader=..., mode_costing=0x7fffc3ffe060, mode=valhalla::sif::TravelMode::kBicycle,      
    options=...) at ../src/thor/bidirectional_astar.cc:628                                                                                                                                     
#6  0x000055555571b51f in valhalla::thor::thor_worker_t::get_path (this=0x7fffc3ffdf20,                                                                                                        
    path_algorithm=0x7fffc3ffe1b0, origin=..., destination=..., costing="bicycle", options=...)                                                                                                
    at ../src/thor/route_action.cc:330                                                         
#7  0x000055555571c98e in valhalla::thor::thor_worker_t::path_depart_at (this=0x7fffc3ffdf20, api=...,                                                                                         
    costing="bicycle") at ../src/thor/route_action.cc:496                                      
#8  0x000055555571acab in valhalla::thor::thor_worker_t::route (this=0x7fffc3ffdf20, request=...)                                                                                              
    at ../src/thor/route_action.cc:258                                                         
#9  0x00005555556f4899 in valhalla::thor::thor_worker_t::work(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&) (this=0x7fffc3f
fdf20,                                         
    job=std::__cxx11::list = {...}, request_info=0x7fffb0038128, interrupt_function=...)                                                                                                       
    at ../src/thor/worker.cc:136                                                               
#10 0x0000555555703097 in std::__invoke_impl<prime_server::worker_t::result_t, prime_server::worker_t::result_t (valhalla::thor::thor_worker_t::*&)(std::__cxx11::list<zmq::message_t, std::all
ocator<zmq::message_t> > const&, void*, std::function<void ()> const&), valhalla::thor::thor_worker_t&, std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std:
:function<void ()>&>(std::__invoke_memfun_ref, prime_server::worker_t::result_t (valhalla::thor::thor_worker_t::*&)(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&,
 void*, std::function<void ()> const&), valhalla::thor::thor_worker_t&, std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*&&, std::function<void ()>&) (
    __f=@0x7fffb0015a20: &virtual valhalla::thor::thor_worker_t::work(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&), __t=..
.,                                             
    __args#0=std::__cxx11::list = {...}, __args#1=@0x7fffc3ffda40: 0x7fffb0038128, __args#2=...)                                                                                               
    at /usr/include/c++/9.2.0/bits/invoke.h:66                                                 
#11 0x0000555555702ae1 in std::__invoke<prime_server::worker_t::result_t (valhalla::thor::thor_worker_t::*&)(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*,
 std::function<void ()> const&), valhalla::thor::thor_worker_t&, std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()>&>(prime_server::work
er_t::result_t (valhalla::thor::thor_worker_t::*&)(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&), valhalla::thor::thor_work
er_t&, std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*&&, std::function<void ()>&) (
    __fn=@0x7fffb0015a20: &virtual valhalla::thor::thor_worker_t::work(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&), __arg
s#0=...,                                       
    __args#1=std::__cxx11::list = {...}, __args#2=@0x7fffc3ffda40: 0x7fffb0038128, __args#3=...)                                                                                               
    at /usr/include/c++/9.2.0/bits/invoke.h:96                                                 
#12 0x0000555555702185 in std::_Bind<prime_server::worker_t::result_t (valhalla::thor::thor_worker_t::*(std::reference_wrapper<valhalla::thor::thor_worker_t>, std::_Placeholder<1>, std::_Plac
eholder<2>, std::_Placeholder<3>))(std::__cxx11::list<zmq::message_t, std::allocator<zmq::message_t> > const&, void*, std::function<void ()> const&)>::__call<prime_server::worker_t::result_t
```

**Comment by PureW** [comment]
config
```
{
  "additional_data": {
    "elevation": "/data/valhalla/elevation/"
  },
  "httpd": {
    "service": {
      "interrupt": "ipc:///tmp/interrupt",
      "listen": "tcp://*:8002",
      "loopback": "ipc:///tmp/loopback"
    }
  },
  "loki": {
    "actions": [
      "locate",
      "route",
      "height",
      "sources_to_targets",
      "optimized_route",
      "isochrone",
      "trace_route",
      "trace_attributes",
      "transit_available"
    ],
    "logging": {
      "color": true,
      "file_name": "path_to_some_file.log",
      "long_request": 100.0,
      "type": "std_out"
    },
    "service": {
      "proxy": "ipc:///tmp/loki"
    },
    "service_defaults": {
      "heading_tolerance": 60,
      "minimum_reachability": 50,
      "node_snap_tolerance": 5,
      "radius": 0,
      "search_cutoff": 35000,
      "street_side_tolerance": 5
    },
    "use_connectivity": true
  },
  "meili": {
    "auto": {
      "search_radius": 50,
      "turn_penalty_factor": 200
    },
    "bicycle": {
      "turn_penalty_factor": 140
    },
    "customizable": [
      "mode",
      "search_radius",
      "turn_penalty_factor",
      "gps_accuracy",
      "interpolation_distance",
      "sigma_z",
      "beta",
      "max_route_distance_factor",
      "max_route_time_factor"
    ],
    "default": {
      "beta": 3,
      "breakage_distance": 2000,
      "geometry": false,
      "gps_accuracy": 5.0,
      "interpolation_distance": 10,
      "max_route_distance_factor": 5,
      "max_route_time_factor": 5,
      "max_search_radius": 100,
      "route": true,
      "search_radius": 50,
      "sigma_z": 4.07,
      "turn_penalty_factor": 0
    },
    "grid": {
      "cache_size": 100240,
      "size": 500
    },
    "logging": {
      "color": true,
      "file_name": "path_to_some_file.log",
      "type": "std_out"
    },
    "mode": "auto",
    "multimodal": {
      "turn_penalty_factor": 70
    },
    "pedestrian": {
      "search_radius": 50,
      "turn_penalty_factor": 100
    },
    "service": {
      "proxy": "ipc:///tmp/meili"
    },
    "verbose": false
  },
  "mjolnir": {
    "admin": "/home/anders/data/admin_2019_10_26-03_00_00.sqlite",
    "global_synchronized_cache": true,
    "hierarchy": true,
    "import_bike_share_stations": false,
    "include_bicycle": true,
    "include_driveways": true,
    "include_driving": true,
    "include_pedestrian": true,
    "logging": {
      "color": true,
      "file_name": "path_to_some_file.log",
      "type": "std_out"
    },
    "max_cache_size": 1000000000,
    "shortcuts": true,
    "tile_dir": "/home/anders/data/planet_tiles/",
    "timezone": "/home/anders/data/tz_world_2019_10_26-03_00_00.sqlite",
    "transit_dir": "/data/valhalla/transit"
  },
  "odin": {
    "logging": {
      "color": true,
      "file_name": "path_to_some_file.log",
      "type": "std_out"
    },
    "service": {
      "proxy": "ipc:///tmp/odin"
    }
  },
  "service_limits": {
    "auto": {
      "max_distance": 5000000.0,
      "max_locations": 20,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    },
    "auto_shorter": {
      "max_distance": 5000000.0,
      "max_locations": 20,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    },
    "bicycle": {
      "max_distance": 500000.0,
      "max_locations": 50,
      "max_matrix_distance": 200000.0,
      "max_matrix_locations": 50
    },
    "bus": {
      "max_distance": 5000000.0,
      "max_locations": 50,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    },
    "hov": {
      "max_distance": 5000000.0,
      "max_locations": 20,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    },
    "isochrone": {
      "max_contours": 4,
      "max_distance": 25000.0,
      "max_locations": 1,
      "max_time": 120
    },
    "max_alternates": 2,
    "max_avoid_locations": 50,
    "max_radius": 200,
    "max_reachability": 100,
    "max_timedep_distance": 500000,
    "motor_scooter": {
      "max_distance": 500000.0,
      "max_locations": 50,
      "max_matrix_distance": 200000.0,
      "max_matrix_locations": 50
    },
    "motorcycle": {
      "max_distance": 500000.0,
      "max_locations": 50,
      "max_matrix_distance": 200000.0,
      "max_matrix_locations": 50
    },
    "multimodal": {
      "max_distance": 500000.0,
      "max_locations": 50,
      "max_matrix_distance": 0.0,
      "max_matrix_locations": 0
    },
    "pedestrian": {
      "max_distance": 250000.0,
      "max_locations": 50,
      "max_matrix_distance": 200000.0,
      "max_matrix_locations": 50,
      "max_transit_walking_distance": 10000,
      "min_transit_walking_distance": 1
    },
    "skadi": {
      "max_shape": 750000,
      "min_resample": 10.0
    },
    "taxi": {
      "max_distance": 5000000.0,
      "max_locations": 20,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    },
    "trace": {
      "max_best_paths": 4,
      "max_best_paths_shape": 100,
      "max_distance": 200000.0,
      "max_gps_accuracy": 100.0,
      "max_search_radius": 100.0,
      "max_shape": 16000
    },
    "transit": {
      "max_distance": 500000.0,
      "max_locations": 50,
      "max_matrix_distance": 200000.0,
      "max_matrix_locations": 50
    },
    "truck": {
      "max_distance": 5000000.0,
      "max_locations": 20,
      "max_matrix_distance": 400000.0,
      "max_matrix_locations": 50
    }
  },
  "thor": {
    "logging": {
      "color": true,
      "file_name": "path_to_some_file.log",
      "long_request": 110.0,
      "type": "std_out"
    },
    "service": {
      "proxy": "ipc:///tmp/thor"
    },
    "source_to_target_algorithm": "select_optimal"
  }
}
```

**Comment by mskurydin** [comment]
Just an idea - If graphreader is used from multiple threads (with global cache) and one of the threads does Clear on graphreader (which removes all the tiles from cache), some other threads may still be holding on raw pointers. In current design cache owns the tiles it gives raw pointers to them to the users.

Discussed this potential issue with Kevin recently (in the context in which valhalla is used in nav-sdk).
It's probably not a big risk for server use case at the moment. We typically call Clear/Trim only when overcommitted, which should be extremely unlikely when using the config with memory mapped tar extract. The tile size in this case does not include the actual tile data and is hardcoded to 1K (the default memory cache size is 1GB). However, for tile directory use case it becomes real (typical tile sizes range from 2 to 10 MB).

In order to properly solve it we should probably return something similar to shared_ptr from graphreader.

**Comment by averkhaturau** [comment]
This issue should be re-checked and probably closed.

**Comment by ktnr** [comment]
I can confirm issues when using `global_synchronized_cache = true`. The errors occur regardless of the number of valhalla threads and when issues request in parallel. I get different errors:
- free(): invalid pointer
- double free or corruption (fasttop)
- malloc(): unaligned tcache chunk detected
- Segmentation fault
- malloc_consolidate(): unaligned fastbin chunk detected

Here is a script to reproduce:
```
from routingpy import Valhalla
from concurrent.futures import ThreadPoolExecutor, as_completed
import itertools

locations = [[7.477554, 46.939949], [8.523617, 46.735583], [8.958152, 46.010689], [7.78884, 47.545814]]
client = Valhalla(base_url='http://localhost:8002')

def fetch_matrix(client, locations, profile="truck", preference="fastest", dry_run=False):
    try:
        clientMatrix = client.matrix(locations=locations, profile=profile, preference=preference, dry_run=dry_run)
        return clientMatrix
    except Exception as e:
        print(e)
        return None

with ThreadPoolExecutor(max_workers=2) as executor: # No errors with max_workers=1
    futures = [executor.submit(fetch_matrix, client, locations, "truck", "fastest", False) for _ in range(5)]
    for future in as_completed(futures):
        result = future.result()
        if result is not None:
            print("Request succeeded")
        else:
            print("Request failed")
```

Tested on 3.4.0 (cbabe7cfbad2225090e4c006c778f1f0a7b3ec4e) built from source.

**Comment by ktnr** [comment]
Thanks for the fast response. Just to confirm that I understand correctly: the high level interface you are talking about includes `valhalla_service`? 

So, I run `./build/valhalla_service valhalla-default.json 1` with `global_synchronized_cache  = true` in the `valhalla.json` config. In this case `global_synchronized_cache  = true` is not expected to work because `valhalla_service` is not thread safe? 

Regarding `routingpy`, I don't see how it would be relevant, as I am using it only the client side to send HTTP requests to valhalla. Anyway, I also get those errors when I run simple curl commands instead of using `routingpy`.  Test script with curl:
<details>

```
import subprocess
import json
from concurrent.futures import ThreadPoolExecutor, as_completed

locations = [[7.477554, 46.939949], [8.523617, 46.735583], [8.958152, 46.010689], [7.78884, 47.545814]]
base_url = 'http://localhost:8002'

def fetch_matrix(locations, profile="truck", preference="fastest"):
	# Construct the payload
	payload = {
		"sources": [{"lon": loc[0], "lat": loc[1]} for loc in locations],
		"targets": [{"lon": loc[0], "lat": loc[1]} for loc in locations],
		"costing": profile
	}
	
	curl_cmd = [
		"curl", "-X", "POST",
		f"{base_url}/sources_to_targets",
		"-H", "Content-Type: application/json",
		"-d", json.dumps(payload)
	]
	
	try:
		result = subprocess.run(curl_cmd, capture_output=True, text=True, check=True)
		response = json.loads(result.stdout)
		return response
	except subprocess.CalledProcessError as e:
		print(e)
		return None

with ThreadPoolExecutor(max_workers=2) as executor:
	futures = [executor.submit(fetch_matrix, locations, "truck", "fastest") for _ in range(5)]
	for future in as_completed(futures):
		result = future.result()
		if result is not None:
			print("Request succeeded")
		else:
			print("Request failed")
```

</details>

You recommend memory mapped tar instead. This just involves building `valhalla_tiles.tar` and setting `tile_extract` in the valhalla config, right?

**Comment by chrstnbwnkl** [comment]
Yeah I think there's a misunderstanding, the library that Nils meant to include is a different one (python bindings that use pybind to call libvalhalla under the hood), routingpy is indeed just an HTTP wrapper

**Comment by kevinkreiser** [comment]
ah yep apologies about that. routingpy, as i was saying, is foreign to me. i was guessing that it wrapped our python bindings (the part thats actually built in this repo). our python bindings and the rest of our high level objects (actor) are not thread safe.

but yeah @chrstnbwnkl knows that project more than i so if its just an http client then you can totally ignore that part of what i was saying.

as for the issue, let me try again :smile: now that i know you are using `valhalla_service`. it doesnt really make sense to call that executable thread safe or not. it should work equally well with either the memory mapped tar (yes tile_extract) or using the tile_dir mode with or without global synchronized cache. if you are getting random memory errors then we have a bug. in the mean time you should be using tile_extract mode as its the fastest mode.

one of use will have to see how the shared memory cache mode regressed. this mode is primarily targeted at use-cases like single user devices where you'd like to do more than one computation at a time (in car navigation systems or android/ios apps)

**Comment by kevinkreiser** [comment]
ah and i just remembered! if you want to use global singleton cache you have to compile with: `-DENABLE_THREAD_SAFE_TILE_REF_COUNT=On`

**Comment by ktnr** [comment]
Thanks! With that flag enabled, I encounter no errors.
